librz/type/path.c

```diff
@@ -122,91 +122,140 @@ RZ_API st64 rz_type_offset_by_path(const RzTypeDB *typedb, RZ_NONNULL const char
+			break;
+		}
+		ut64 idx_offset = offset - idx * stride;
+		char *newpath = rz_str_newf("%s[%" PFMT64u "]", prefix, idx);
+		if (idx_offset == 0) {
+			RzTypePath *tpath = rz_type_path_new(type->array.type, newpath);
+			if (tpath) {
+				rz_list_append(list, tpath);
+			}
+		}
+		collect_type_paths(typedb, list, type->array.type, newpath, idx_offset, depth - 1);
+		if (idx_offset != 0) {
 			free(newpath);
 		}
+		break;
+	}
+	default:
+		break;
 	}
-	return result;
 }
 
 /**
- * \brief Returns the list of all type paths matching the offset
+ * \brief Returns the list of all paths into the base type matching the offset
  *
- * \param typedb Types Database instance
- * \param btype The base type
- * \param offset The offset of the path to match against
+ * \param btype The root type from which paths are searched
+ * \param offset The offset into \p btype to match against
+ * \param max_depth Maximum number of components a path may have
  */
-RZ_API RZ_OWN RzList /*<RzTypePath *>*/ *rz_type_path_by_offset(const RzTypeDB *typedb, RzBaseType *btype, ut64 offset) {
+RZ_API RZ_OWN RzList /*<RzTypePath *>*/ *rz_base_type_path_by_offset(const RzTypeDB *typedb, RzBaseType *btype, ut64 offset, unsigned int max_depth) {
+	rz_return_val_if_fail(typedb && btype && btype->name, NULL);
 	RzList *list = rz_list_newf((RzListFree)rz_type_path_free);
-	if (btype->kind == RZ_BASE_TYPE_KIND_STRUCT) {
-		RzType *t = rz_type_identifier_of_base_type(typedb, btype, false);
-		RzTypeStructMember *memb;
-		ut64 memb_offset = 0;
-		rz_vector_foreach(&btype->struct_data.members, memb) {
-			if (memb_offset == offset) {
-				RzTypePath *tpath = rz_type_path_new(t, rz_str_newf("%s.%s", btype->name, memb->name));
-				if (tpath) {
-					rz_list_append(list, tpath);
-				}
-			}
-			// We go into the nested structures/unions if they are members of the structure
-			char *path = rz_str_newf("%s.%s", btype->name, memb->name);
-			structured_member_walker(typedb, list, t, memb->type, path, memb_offset + offset);
-			memb_offset += rz_type_db_get_bitsize(typedb, memb->type) / 8;
-			free(path);
-		}
-	} else if (btype->kind == RZ_BASE_TYPE_KIND_UNION) {
-		// This function makes sense only for structures since union
-		// members have exact same offset
-		// But if the union has compound members, e.g. structures, their
-		// internal offsets can be different
-		RzType *t = rz_type_identifier_of_base_type(typedb, btype, false);
-		RzTypeUnionMember *memb;
-		rz_vector_foreach(&btype->union_data.members, memb) {
-			char *path = rz_str_newf("%s.%s", btype->name, memb->name);
-			structured_member_walker(typedb, list, t, memb->type, path, offset);
-			free(path);
-		}
-	} else {
-		rz_warn_if_reached();
+	if (!list) {
+		return NULL;
+	}
+	collect_base_type_paths(typedb, list, btype, "", offset, max_depth);
+	return list;
+}
+
+/**
+ * \brief Returns the list of all paths into the type matching the offset
+ *
+ * \param type The root type from which paths are searched
```

Comment:
Also missing `typedb` parameter?

Reply:
And also `RzType *type` is const?

Reply:
same as above, there isn't really anything valuable to document about the `typedb`.